runtime/src/tokio.rs

```diff
@@ -267,7 +270,19 @@ impl crate::Runner for Runner {
         F: Future + Send + 'static,
         F::Output: Send + 'static,
     {
-        self.executor.runtime.block_on(f)
+        // Run the provided task
+        let result = {
+            let runtime = self.executor.runtime.read().unwrap();
+            runtime.as_ref().unwrap().block_on(f)
+        };
+
+        // Shutdown the runtime after the task completes
+        //
+        // Dropping the runtime will result in a blocking shutdown of all outstanding
+        // tasks. If we used `.shutdown_background()`, we would not wait for all tasks
+        // to halt (and may leak those resources).
+        let _ = self.executor.runtime.write().unwrap().take().unwrap();
```

Comment:
Looks like explicitly waiting for all tasks to complete means some tests no longer exit.

Before merging, we should confirm examples don't have this issue.

Reply:
Specifically `p2p::authenticated::test_tokio_connectivity` fails (likely because we never call shutdown on `p2p`).